x-pack/legacy/plugins/transform/public/shim.ts

```diff
@@ -0,0 +1,149 @@
+import { I18nContext } from 'ui/i18n';
+
+import chrome from 'ui/chrome';
+import { DOC_LINK_VERSION, ELASTIC_WEBSITE_URL } from 'ui/documentation_links';
+import { management, MANAGEMENT_BREADCRUMB } from 'ui/management';
+import { fatalError, toastNotifications } from 'ui/notify';
+import routes from 'ui/routes';
+import { docTitle } from 'ui/doc_title/doc_title';
+
+import { HashRouter } from 'react-router-dom';
+
+// @ts-ignore: allow traversal to fail on x-pack build
+import { createUiStatsReporter } from '../../../../../src/legacy/core_plugins/ui_metric/public';
+import { SavedSearchLoader } from '../../../../../src/legacy/core_plugins/kibana/public/discover/types';
+
+export interface AppCore {
+  chrome: typeof chrome;
+  i18n: {
+    [i18nPackage: string]: any;
+    Context: typeof I18nContext;
+    FormattedMessage: typeof FormattedMessage;
+    FormattedDate: typeof FormattedDate;
+    FormattedTime: typeof FormattedTime;
+  };
+  notification: {
+    fatalError: typeof fatalError;
+    toastNotifications: typeof toastNotifications;
+  };
+  savedSearches: {
+    getClient(): any;
+    setClient(client: any): void;
+  };
+}
+
+export interface AppPlugins {
+  management: {
+    sections: typeof management;
+  };
+}
+
+export interface Core extends AppCore {
+  chrome: typeof chrome;
+  http: {
+    getClient(): any;
+    setClient(client: any): void;
+  };
+  routing: {
+    registerAngularRoute(path: string, config: object): void;
+    registerRouter(router: HashRouter): void;
+    getRouter(): HashRouter | undefined;
+  };
+  documentation: {
+    esDocBasePath: string;
+    esPluginDocBasePath: string;
+    esStackOverviewDocBasePath: string;
+  };
+  docTitle: {
+    change: typeof docTitle.change;
+  };
+}
+
+export interface Plugins extends AppPlugins {
+  management: {
+    sections: typeof management;
+    constants: {
+      BREADCRUMB: typeof MANAGEMENT_BREADCRUMB;
+    };
+  };
+  uiMetric: {
+    createUiStatsReporter: typeof createUiStatsReporter;
+  };
+}
+
+export function createPublicShim(): { core: Core; plugins: Plugins } {
+  // This is an Angular service, which is why we use this provider pattern
+  // to access it within our React app.
+  let httpClient: ng.IHttpService;
+  // This is an Angular service, which is why we use this provider pattern
+  // to access it within our React app.
+  let savedSearches: SavedSearchLoader;
+
+  let reactRouter: HashRouter | undefined;
+
+  return {
+    core: {
+      i18n: {
+        ...i18n,
+        Context: I18nContext,
+        FormattedMessage,
+        FormattedDate,
+        FormattedTime,
+      },
+      routing: {
+        registerAngularRoute: (path: string, config: object): void => {
+          routes.when(path, config);
+        },
+        registerRouter: (router: HashRouter): void => {
+          reactRouter = router;
+        },
+        getRouter: (): HashRouter | undefined => {
+          return reactRouter;
+        },
```

Comment:
FYI routing will not be a Core concern so there will never be an API in core like this. Instead you can use the `core.application.register` API to mount your application and setup your own router.

Reply:
I've removed the code related to React's router here since it's done via direct imports in the rest of the code anyway. I'm leaving the `registerAngularRoute()` in as is for now, I'm not sure yet how to use `core.application.register` so I'll leave that for another time. Added a note in #45352.